### PR TITLE
Fix unicode issues in HttpError and BatchHttpRequest

### PR DESCRIPTION
**Pull request by methane**
1) HttpError parses http response by json.loads. It accepts only
unicode on PY3 while response is bytes. So decode response first.

2) BatchHttpRequest uses email.parser.FeedParser. It accepts only
unicode on PY3, too. So encode parsed response to emulate normal
http response.

### CODE REVIEW

**Comment by nathanielmanistaatgoogle** [comment]
(1) Squash commits?
(2) Write a longer, more detailed commit message? (http://chris.beams.io/posts/git-commit/#seven-rules)
(3) Why does Travis CI show test failure for this change?

File: googleapiclient/errors.py [hunk 1]
```diff
@@ -44,9 +44,9 @@ def _get_reason(self):
     """Calculate the reason for the error from the response content."""
     reason = self.resp.reason
     try:
-      data = json.loads(self.content)
+      data = json.loads(self.content.decode('utf-8'))
```
Comment (pferate) [inline googleapiclient/errors.py]: I think that this should not be included within this block, it should be in a separate one before this.

If there is actually a 'reason' but it fails on UnicodeError, it will seem as if there is no reason when in fact there is.

File: googleapiclient/errors.py [hunk 2]
```diff
@@ -44,9 +44,9 @@ def _get_reason(self):
...
       reason = data['error']['message']
-    except (ValueError, KeyError):
+    except (ValueError, KeyError, UnicodeError):
```
Comment (pferate) [inline googleapiclient/errors.py]: It should be AttributeError (See output of nosetests/Travis)

**Comment by pferate** [comment]
@nathanielmanistaatgoogle 
1) I was going to say the same, I've fought that battle a few times here recently
2) I like that post, what do you think about linking it somewhere in the README?
3) The wrong Error was being caught, I pointed it out in the line notes.

File: googleapiclient/http.py [hunk 3]
```diff
@@ -1252,11 +1252,10 @@ def _execute(self, http, order, requests):
     if resp.status >= 300:
       raise HttpError(resp, content, uri=self._batch_uri)
 
-    # Now break out the individual responses and store each one.
```
Comment (pferate) [inline googleapiclient/http.py]: I agree that this shouldn't be here, but it is not related to the rest of the commit/PR; move it to a separate commit/PR.

**Comment by pferate** [comment]
There are 3 ways that you have for getting to the decode/encode calls:
1) try/except
2) if not six.PY2:
3) if isinstance(content, six.text_type):

Personally, I like "It's better to ask forgiveness than permission", so I would go with number 1.  I'm not sure if @nathanielmanistaatgoogle has a preference.

File: googleapiclient/errors.py (hunk 1 above)
Comment (methane) [inline googleapiclient/errors.py]: http response should bytes always. It should have `.decode()` on both of PY2 and PY3.
AttributeError is just a bug. Catching it is bad idea.

File: googleapiclient/http.py (hunk 3 above)
Comment (methane) [inline googleapiclient/http.py]: done

**Comment by methane** [comment]
> (1) Squash commits?
Done

> (2) Write a longer, more detailed commit message? (http://chris.beams.io/posts/git-commit/#seven-rules)
Done, bug I'm not a good English writer.

> (3) Why does Travis CI show test failure for this change?
There were bugs remained relating BatchHttpRequest. I've fixed it.

File: googleapiclient/http.py [hunk 4]
```diff
@@ -1257,6 +1257,8 @@ def _execute(self, http, order, requests):
 
     # Prepend with a content-type header so FeedParser can handle it.
     header = 'content-type: %s\r\n\r\n' % resp['content-type']
+    if not six.PY2:
+      content = content.decode('utf-8')
```
Comment (methane) [inline googleapiclient/http.py]: May content have binary data?
If so, [surrogate escape](https://www.python.org/dev/peps/pep-0383/) may help us.
But I'm not sure about how FeedParser treat surrogate escaped text.

File: googleapiclient/errors.py (hunk 1 above)
Comment (methane) [inline googleapiclient/errors.py]: I've added assert in `__init__` to catch bugs early.

Comment (pferate) [inline googleapiclient/errors.py]: Looking at it again, I think your idea is better.  What do you think about checking for bytes and raising an error instead of using the assert?

Comment (methane) [inline googleapiclient/errors.py]: > Looking at it again, I think your idea is better. What do you think about checking for bytes and raising an error instead of using the assert?

I've did it. Raise TypeError instead of AssertionError

File: googleapiclient/errors.py [hunk 5]
```diff
@@ -37,6 +37,9 @@ class HttpError(Error):
   @util.positional(3)
   def __init__(self, resp, content, uri=None):
     self.resp = resp
+    if not isinstance(content, bytes):
+        # When see this exception in test, testdata may be wrong
```
Comment (nathanielmanistaatgoogle) [inline googleapiclient/errors.py]: Drop this comment; test data passing through this function should be all corrected in this commit.

Comment (methane) [inline googleapiclient/errors.py]: done

**Comment by pferate** [comment]
LGTM.

**Comment by nathanielmanistaatgoogle** [comment]
Thanks for the contribution. :-)
